[DataPipe] adding/removing __len__ for different DataPipe

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #64404
* __->__ #64398



cc @VitalyFedyunin @ejguan

Differential Revision: [D30710437](https://our.internmc.facebook.com/intern/diff/D30710437)

NivekT committed:
adding/removing __len__ for different DataPipes

[ghstack-poisoned]

NivekT commented on torch/utils/data/datapipes/iter/selecting.py:
```diff
@@ -76,7 +76,3 @@ def _isNonEmpty(self, data):
         return r
-
-
-    def __len__(self):
```
This is being removed because as @ejguan mentioned:

"With `__len__` implemented, `isinstance(dp, Sized)` is always True. I don't think it's the expected behavior."

NivekT replied:
I'm undoing this and moving it to a different PR.

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30710437).

ejguan commented on test/test_datapipe.py:
```diff
@@ -1626,6 +1626,28 @@ def test_simple_sharding(self):
 
         self.assertEqual(sorted(all_items), sorted(items))
 
+    @skipIfNoDill
```
`skipIfNoDill` is not needed here, right?

NivekT committed:
Update on "[DataPipe] adding/removing __len__ for different DataPipe"




cc @VitalyFedyunin @ejguan

Differential Revision: [D30710437](https://our.internmc.facebook.com/intern/diff/D30710437)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30710437).

ejguan commented on torch/utils/data/datapipes/iter/grouping.py:
```diff
@@ -28,6 +28,12 @@ def __iter__(self):
 
+    def __len__(self):
+        if isinstance(self.source_datapipe, Sized):
+            return len(self.source_datapipe) // self.num_of_instances +\
+                (1 if (self.instance_id < len(self.source_datapipe) % self.num_of_instances) else 0)
```
I guess it's nit: you can do `(len(self.source_datapipe) + self.num_of_instances -1) // self.num_of_instances` to get the same result without ternary op.

NivekT replied:
I thought about it but was worried that it was unintuitive for readers. I will switch to that then.

NivekT replied:
Actually, it needs to be
`(len(self.source_datapipe) + (self.num_of_instances - self.instance_id) - 1) // self.num_of_instances`

So I think it might be better to leave it as it was.

ejguan reviewed (approved):
LGTM

NivekT committed:
Update on "[DataPipe] adding/removing __len__ for different DataPipe"




cc @VitalyFedyunin @ejguan

Differential Revision: [D30710437](https://our.internmc.facebook.com/intern/diff/D30710437)

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] adding/removing __len__ for different DataPipe"




cc @VitalyFedyunin @ejguan

Differential Revision: [D30710437](https://our.internmc.facebook.com/intern/diff/D30710437)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30710437).
